lib/absinthe/phase/schema/validation/unique_field_names.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Absinthe.Phase.Schema.Validation.UniqueFieldNames do
+  @moduledoc false
+
+  @behaviour Absinthe.Phase
+  alias Absinthe.Blueprint
+
+  def run(bp, _) do
+    bp =
+      bp
+      |> Blueprint.prewalk(&handle_schemas(&1, :name))
+
+    {:ok, bp}
+  end
+
+  defp handle_schemas(%Blueprint.Schema.SchemaDefinition{} = schema, key) do
+    schema = Blueprint.prewalk(schema, &validate_types(&1, key))
+    {:halt, schema}
+  end
+
+  defp handle_schemas(obj, _) do
+    obj
+  end
+
+  defp validate_types(%type{} = object, key)
+       when type in [
+              Blueprint.Schema.InputObjectTypeDefinition,
+              Blueprint.Schema.InterfaceTypeDefinition,
+              Blueprint.Schema.ObjectTypeDefinition
+            ] do
+    fields =
+      for field <- object.fields do
+        if duplicate?(object.fields, field, key) do
```

Comment:
I think the first arg here should be a Map of the `object.field` names, otherwise this becomes `O(n^2)`. We can do:

```
name_counts = Enum. frequencies_by(object.fields, &Map.get(&1, key))

def duplicate?(name_counts, field, key) do
  field_identifier = Map.get(field, key)
  Map.get(name_counts, field_identifier, 0) > 1
end
```

Reply:
Done :)

I copied the duplicate approach from other parts of the code. E.g. `Absinthe.Phase.Document.Validation.UniqueVariableNames`. These could probably also be optimized, esp since they are run in the document pipeline and not in the schema pipeline, so the cost is incurred for every request.

Reply:
Ah that sounds like a great followup!

Reply:
For now this PR can be merged